File: examples/startup/call_transfer.py
"""
Call Transfer
=============
Dynamic call transfer and hangup using CallControl. The on_message handler
receives a CallControl object as its second argument, enabling mid-call
actions like transferring to a human agent or ending the conversation.

Requirements:
    pip install patter python-dotenv

Environment variables (.env):
    OPENAI_API_KEY      - OpenAI API key
    TWILIO_ACCOUNT_SID  - Twilio account SID
    TWILIO_AUTH_TOKEN   - Twilio auth token
    TWILIO_PHONE_NUMBER - Your Twilio phone number (E.164 format)
    WEBHOOK_URL         - Public URL where Twilio can reach this server
"""

import asyncio
import os

from dotenv import load_dotenv

load_dotenv()

from patter import Patter

HUMAN_AGENT_NUMBER = "+15550001111"

# ── Initialise ───────────────────────────────────────────────────────
phone = Patter(
    mode="local",
    openai_key=os.getenv("OPENAI_API_KEY"),
    twilio_sid=os.getenv("TWILIO_ACCOUNT_SID"),
    twilio_token=os.getenv("TWILIO_AUTH_TOKEN"),
    phone_number=os.getenv("TWILIO_PHONE_NUMBER"),
    webhook_url=os.getenv("WEBHOOK_URL"),
)

# ── Agent ────────────────────────────────────────────────────────────
agent = phone.agent(
    system_prompt=(
        "You are a friendly first-line support agent. Try to resolve the "
        "caller's issue yourself. If they ask to speak to a human or a "
        "manager, let them know you're transferring them. If they say "
        "goodbye, wish them well and end the call."
    ),
    voice="alloy",
    first_message="Hi there! I'm here to help — what can I do for you?",
)


# ── Message handler with call control ────────────────────────────────
async def on_message(data: dict, call_control) -> str | None:
    """Handle user messages with the ability to transfer or hang up."""
    user_text = data.get("text", "").lower()

    # Check if the call has already ended (transfer or hangup)
    if call_control.ended:
        return None

    # Transfer to a human agent
    if any(phrase in user_text for phrase in ("speak to human", "speak to a person", "transfer me", "talk to a manager")):
        await call_control.transfer(HUMAN_AGENT_NUMBER)
        return "Sure, let me transfer you to a team member right now."

    # End the call gracefully
    if any(phrase in user_text for phrase in ("goodbye", "bye", "that's all", "hang up")):
        await call_control.hangup()
        return "Thanks for calling! Have a great day."

    # Fall back to the LLM for everything else
    return None


# ── Start the server ─────────────────────────────────────────────────
if __name__ == "__main__":
    asyncio.run(phone.serve(agent, port=8000, on_message=on_message))

File: examples/startup/call_transfer.ts
/**
 * @file call_transfer.ts
 * @description Call control with live transfer and hangup.
 *
 * Shows how to use the callControl parameter in onMessage to transfer
 * callers to a human agent or gracefully end the call based on
 * conversation context. The agent triages incoming calls and escalates
 * when it cannot resolve the issue.
 *
 * Prerequisites:
 *   - OPENAI_API_KEY, TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN in .env
 *   - A Twilio phone number pointed at your webhook URL
 *
 * Usage:
 *   npx ts-node call_transfer.ts
 */

import { Patter } from "patter";
import dotenv from "dotenv";

dotenv.config();

const HUMAN_AGENT_NUMBER = process.env.TRANSFER_NUMBER ?? "+15551234567";

const phone = new Patter({
  mode: "local",
  openaiKey: process.env.OPENAI_API_KEY!,
  twilioSid: process.env.TWILIO_ACCOUNT_SID!,
  twilioToken: process.env.TWILIO_AUTH_TOKEN!,
  phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
  webhookUrl: process.env.WEBHOOK_URL!,
});

const agent = phone.agent({
  systemPrompt: `You are a triage assistant for an insurance company. Help callers
with simple questions about their policy. If the caller needs to file a claim
or requests to speak with a human, let them know you are transferring them.
If the caller says "goodbye" or the issue is resolved, end the call politely.`,
  firstMessage: "Hello! Welcome to SafeGuard Insurance. How can I assist you today?",
});

agent.onMessage((message, callControl) => {
  console.log(`[${message.role}] ${message.content}`);

  if (callControl.ended) {
    console.log("Call has already ended, skipping control logic.");
    return;
  }

  const text = message.content.toLowerCase();

  if (text.includes("transferring you") || text.includes("connect you to a human")) {
    console.log(`Transferring call to ${HUMAN_AGENT_NUMBER}...`);
    callControl.transfer(HUMAN_AGENT_NUMBER);
    return;
  }

  if (text.includes("goodbye") || text.includes("have a great day")) {
    console.log("Ending call gracefully.");
    callControl.hangup();
  }
});

async function main(): Promise<void> {
  await phone.serve({ agent, port: 8000 });
  console.log("Call-transfer agent listening on port 8000");
}

main();
